py3status/modules/do_not_disturb.py

```diff
@@ -1,104 +1,225 @@
+        if not self.server:
+            for server in servers:
+                if server:
+                    try:
+                        if self.py3.command_output(['pgrep', '-f', server]):
+                            self.server = server
+                            break
+                    except self.py3.CommandError:
+                        pass
+            else:
+                self.server = self.py3.check_commands(servers[:-1]) or 'dunst'
+        elif self.server not in servers:
+            raise Exception(STRING_INVALID_SERVER.format(self.server))
+        else:
+            command = self.server.replace('notifyd', 'notifyd-config')
+            if not self.py3.check_commands(command):
+                raise Exception(STRING_NOT_INSTALLED.format(command))
+
+        if self.server == 'dunst':
+            self.backend = Dunst(self)
+        elif self.server == 'xfce4-notifyd':
+            self.backend = Xfce4_notifyd(self)
+        elif self.server == 'lxqt-notificationd':
+            self.backend = Lxqt_notificationd(self)
+
+        if self.state is not None:
+            if self.state == 'last':
+                self.state = self.py3.storage_get('state') or 0
+            if self.state in [False, True]:
+                self.backend.toggle(self.state)
+            else:
+                raise Exception(STRING_INVALID_STATE.format(self.state))
 
     def do_not_disturb(self):
-        state = self.state_on if self.is_on else self.state_off
-        response = {
-            'cached_until': self.py3.CACHE_FOREVER,
-            'full_text': self.py3.safe_format(self.format, {'state': state}),
-            'color': self.py3.COLOR_BAD if self.is_on else self.py3.COLOR_GOOD
+        self.state = self.backend.get_state()
+
+        return {
+            'cached_until': self.py3.time_in(self.cache_timeout),
+            'full_text': self.py3.safe_format(
+                self.format, {
+                    'state': int(self.state),
+                    'server': self.server,
+                }
+            )
         }
-        return response
 
     def on_click(self, event):
-        if event['button'] == 1:
-            self.is_on = not self.is_on
-            if self._is_dunst():
-                new_flag = "--signal {}".format(self.dunst_signal_on
-                                                if self.is_on else self.dunst_signal_off)
-                system("killall {} dunst".format(new_flag))
-            else:
-                if self.is_on:
-                    self.running.set()
-                else:
-                    self.running.clear()
+        self.state = not self.state
+        self.py3.storage_set('state', self.state)
+        self.backend.toggle(self.state)
 
     def kill(self):
-        if not self._is_dunst():
-            self.killed.set()
-            self.running.set()  # ensure the thread won't hang
+        self.backend.kill()
```

Comment:
can we not just have a daemon thread?

Reply:
Not familiar with this. I'll have to try it for this and probably `usbguard` too.